### PR TITLE
Add rotary and use lane types

### PR DESCRIPTION
**Pull request by freenerd**
These two types were not present yet.

@1ec5 What other changes are required to bring these types in?

### CODE REVIEW

**Comment by 1ec5** [comment]
This fixes #79.

**Comment by 1ec5** [comment]
We’ll need to add support for the `rotary_name` type as well. Let’s land #91 first; then we can either add a `junctionNames` property on RouteStep or add a heuristic for rotaries that puts `name` in the `destinations` property and `rotary_name` in the `names` property, for consistency with ramps.

**Comment by freenerd** [comment]
> we can either add a junctionNames property on RouteStep 
> add a heuristic for rotaries that puts name in the destinations property and rotary_name in the names property, for consistency with ramps.

For integrating osrm-text-instructions it will be more useful to have direct access to the `rotary_name` property. I'd rather not have heuristics that mangle with properties, but instead retain the original response structure in the object and add helper functions that return heuristically-derived results

**Comment by 1ec5** [comment]
> For integrating osrm-text-instructions it will be more useful to have direct access to the rotary_name property.

Direct equivalence to JSON properties is a non-goal for this library, but #60 would keep the original serialized JSON around for anyone who finds it useful. The tradeoff in that case is higher memory usage.

Once #91 lands, the logic I’m proposing for RouteStep would look something like this:

```swift
if maneuverType == .rotary {
    destinations = road.names ?? road.destinations
    names = road.rotaryNames
} else {
    destinations = road.destinations
    names = road.rotaryNames
}
```

I don’t consider this to be “mangling”, since you can easily reverse this logic:

```swift
if maneuverType == .rotary {
    rotaryName = step.destinations.componentsJoined(by: ";")
    name = step.names.componentsJoined(by: ";")
} else {
    destination = step.destinations.componentsJoined(by: ";")
    name = step.names.componentsJoined(by: ";")
}
```

The only problematic case would be for a rotary tagged with `destination`, but the `destination` tag seems ill-suited to rotaries anyways.

You’ll probably have a lot more difficulty reversing the logic in #91 that deals with `ref`, but again, the point was never to match the Directions API response one-for-one.

**Comment by freenerd** [comment]
`destinations = road.names ?? road.destinations`

This will be problematic for instructions, since we loose semantic if `destinations` is names or destinations. For the difference, see [this example](https://github.com/Project-OSRM/osrm-text-instructions/blob/53c52c1de0489901e0492ce6d1b129e76c61f9a4/instructions/en.json#L315-L316):

```
                   "name": "Enter {rotary_name} and exit onto {way_name}",
                    "destination": "Enter {rotary_name} and exit towards {destination}"
```

Take this example:

```js
{
  "rotary_name": "Dupont Circle",
  "name":  "6th Avenue",
}
```

If we only expose `destination` then the correct instruction `Enter Dupont Circle and exit onto 6th Avenue` would become `Enter Dupont Circle and exit towards 6th Avenue`.

**Comment by 1ec5** [comment]
> If we only expose `destination` then the correct instruction `Enter Dupont Circle and exit onto 6th Avenue` would become `Enter Dupont Circle and exit towards 6th Avenue`.

As I pointed out in https://github.com/mapbox/MapboxDirections.swift/pull/93#issuecomment-259525631, I think you can safely ignore the destination for rotary instructions.

**Comment by 1ec5** [comment]
An alternative would be to put `rotary_name` into the `name` property and `name` into a new `exitName` property (so named for consistency with `exitIndex`). If we go that route, we should consider putting `name` into `exitName` and leaving `name` empty for roundabout instructions.

**Comment by freenerd** [comment]
@1ec5 This sounds good to me, since all information is carried through and no semantics are lost. To paraphrase:

```swift
if type == .rotary || type == .roundabout {
  step.name = json.rotary_name
  step.destinations = json.destinations
  step.exitName = step.name // empty on .roundabout
}
```

**Comment by 1ec5** [comment]
Yes, that’s the idea. My only reservation is that there’s already a proliferation of road attribute properties on RouteStep. Ideally we’d shunt them all into a struct, but that won’t bridge to Objective-C; meanwhile, an Objective-C class would be overkill. But if it helps us rationalize our naming logic, then so be it.

I’m assuming we don’t want this behavior for `TurnAtRoundabout` since it’s supposed to be similar to an ordinary intersection.

**Comment by 1ec5** [comment]
bcca7e049f3b7146168920fd885b779c10b60777 implements the approach described in https://github.com/mapbox/MapboxDirections.swift/pull/93#issuecomment-259670333. Ready for review.

File: MapboxDirections/MBRouteStep.swift [hunk 1]
```diff
@@ -170,16 +170,32 @@ public enum ManeuverType: Int, CustomStringConvertible {
...
      
-     The exit index indicates the number of roundabout exits up to and including the exit that the user must take.
+     The step’s name is the name of the road to take after exiting the roundabout. The exit index indicates the number of roundabout exits up to and including the exit to take.
      */
     case TakeRoundabout
     
     /**
+     The step requires the user to enter, traverse, and exit a large, named roundabout (traffic circle or rotary).
+     
+     The step’s name is the name of the road to take after exiting the roundabout. The exit index indicates the number of rotary exits up to and including the exit that the user must take.
```
Comment (freenerd) [inline MapboxDirections/MBRouteStep.swift]: I think this is different now with `exitNames`.

> The step's name is the name of the rotary itself. The exit name is the name of the road to take after exiting the roundabout.

**Pull request merged**